Project 9

ttruedell opened the pull request:
Initial implementation of features for the project, including:

- Added functionality for editing user profile and avatar.
- Implemented card addition and deletion with modal interactions.
- Set up API integration for user and card management.
- Ensured error handling and validation for forms.
- Please review the changes in the project-9 branch.

Note: Due to the node_modules  deleted files being tracked from project-9 branch, this made the initial pull request too large to compare with gh-pages.
Therefore, a branch of project-9 -- project-9-backup -- was created. The project-9 branch was hard reset, and project-9-backup was merged into project-9.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,104 @@
+      .then(([user, cards]) => {
+        return { user, cards };
+      })
+      .catch((error) => console.error("Error loading data:", error));
```
It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`    after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,104 @@
+      method: "GET",
+      headers: this.headers,
+    })
+      .then(this._checkResponse)
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/Popup.js:
```diff
@@ -9,6 +9,7 @@ export default class Popup {
     this._popupElement.classList.add("modal_opened");
     document.addEventListener("keydown", this._handleEscClose);
     document.addEventListener("mousedown", this._handleClickOutsideModal);
+    this.resetButtonText(".modal__button");
```
`this.resetButtonText(".modal__button");` should be removed from `Popup` because the image modal doesn't have any button. 

There should **not** be any code in universal functions/classes that doesn’t suit every popup.

gennady-bars commented on src/components/Popup.js:
```diff
@@ -57,7 +58,25 @@ export default class Popup {
+
+  //
+  //
+  setButtonText(buttonSelector, newText) {
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/Popup.js:
```diff
@@ -57,7 +58,25 @@ export default class Popup {
     closeButton.addEventListener("click", () => this.close());
   }
+
+  //
```
`setButtonText` and `resetButtonText` should be removed from `Popup`

There should **not** be any code in universal functions/classes that doesn’t suit every popup.

gennady-bars commented on src/components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,22 @@
+    this._confirmButton.addEventListener("click", (event) => {
+      event.preventDefault();
+      this._handleConfirm();
+      this.setButtonText(".modal__button_delete-card", "Saving...");
```
You have found the button as `this._confirmButton`. No need to search it again inside methods

gennady-bars commented on src/components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,22 @@
+      event.preventDefault();
+      this._handleConfirm();
+      this.setButtonText(".modal__button_delete-card", "Saving...");
+      this.close();
```
You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well



**Please, correct it everywhere**

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -6,6 +6,9 @@ export default class PopupWithForm extends Popup {
     this._handleFormSubmit = handleFormSubmit;
     this._form = this._popupElement.querySelector(".modal__form");
     this._inputList = this._form.querySelectorAll(".modal__input");
+    this._confirmButton = this._popupElement.querySelector(
```
`_confirmButton` should not be here because `PopupWithForm` is for any modal with a form


* There should **not** be any code in universal functions/classes that doesn’t suit every popup.

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -6,6 +6,9 @@ export default class PopupWithForm extends Popup {
+      ".modal__button_confirm-form"
+    );
   }
 
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // Here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -21,12 +24,13 @@ export default class PopupWithForm extends Popup {
     this._form.addEventListener("submit", (event) => {
       event.preventDefault();
       this._handleFormSubmit(this._getInputValues());
+      this.setButtonText(".modal__button_confirm-form", "Saving...");
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -21,12 +24,13 @@ export default class PopupWithForm extends Popup {
       event.preventDefault();
       this._handleFormSubmit(this._getInputValues());
+      this.setButtonText(".modal__button_confirm-form", "Saving...");
       this.close();
```
* You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

gennady-bars commented on src/pages/index.js:
```diff
@@ -23,16 +34,19 @@ const profileDescriptionInput = document.querySelector(
-// Elements: Add Card Modal
+// Elements: Add Card Modal//
 const addNewCardButton = document.querySelector(".profile__add-button");
 const cardForm = document.querySelector("#add-card-form");
```
**Could be improved**

We can get any form using their **unique** attribute `name` or `id` from `document.forms`

```js

const profileForm = document.forms["profile-form"];
const cardForm = document.forms["card-form"];

```

You need to add `name` or `id` first in `index.html` to achieve that.

More info here https://developer.mozilla.org/en-US/docs/Web/API/Document/forms

gennady-bars commented on src/pages/index.js:
```diff
@@ -23,16 +34,19 @@ const profileDescriptionInput = document.querySelector(
+let currentCard;
+
+//Elements: Form Validators//
 const editFormValidator = new FormValidator(settings, profileEditForm);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object: `formValidators`.  Then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // Here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // Here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -41,70 +55,243 @@ addFormValidator.enableValidation();
+  .then(({ user, cards }) => {
+    console.log("Initial user data:", user);
+    console.log("Initial cards data:", cards);
+    const cardElement = cards.map((cardData) => createCard(cardData));
```
`createCard` should be used inside method `renderer`, which should be used inside method `renderItems` according to the task

Your code will be like this


```js
  .then(({ user, cards }) => {
    section.renderItems(cards);
    userInfo.setUserInfo(user);

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -41,70 +55,243 @@ addFormValidator.enableValidation();
+    const userData = {
+      name: data.name,
+      about: data.about,
+    };
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -41,70 +55,243 @@ addFormValidator.enableValidation();
+  },
+});
+
+// function addInitialCards(cards) {
+//   const promises = cards.map((cardData) => {
+//     console.log("Adding card:", cardData);
+//     return api
+//       .addCard(cardData)
```
**Could be improved**

Please, remove **commented code** to clean up the project

gennady-bars commented on src/pages/index.js:
```diff
@@ -41,70 +55,243 @@ addFormValidator.enableValidation();
+    .then(() => {
+      card.removeCard();
+    })
+    .catch((error) => console.error("Error deleting card:", error));
```
Good job that you `catch` possible errors at the end of server requests

gennady-bars commented on src/pages/index.js:
```diff
@@ -41,70 +55,243 @@ addFormValidator.enableValidation();
+      .addCard(cardData)
+      .then((newCard) => {
+        if (newCard && newCard.name && newCard.link) {
+          const cardElement = createCard({
+            name: newCard.name,
+            link: newCard.link,
+            _id: newCard._id,
+          });
```
`newCard` is a ready object for creating

```js

          const cardElement = createCard(newCard);

```

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Tyson!

Great job on submitting your project! This is a significant milestone, and you should be proud of your progress.

**Parts I really liked in your project:**

* Good job that you `catch` possible errors at the end of server requests


There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code   and the comments below:
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`    after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* When I add new cards, they appear with **black** hearts, but they should not be liked initially
* The avatar form is not validated
* The avatar should be **round** to match the design
* `this.resetButtonText(".modal__button");` should be removed from `Popup` because the image modal doesn't have any button. 
* There should **not** be any code in universal functions/classes that doesn’t suit every popup.
* You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all  comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

ttruedell commented:
Refactored Code and Addressed Feedback

- Refactored index.js, Popup.js, PopupWithForm.js, and PopupWithConfirm.js as follows:
- Implemented form validation for the avatar input in PopupWithForm.js.
- Removed unnecessary resetButtonText call in Popup.js related to image modals.
- Ensured all comments in the code are aligned with best practices and project requirements.

Additional Changes:

- Removed catch blocks from Api.js to ensure errors are handled uniformly at the end of requests in index.js.
- Adjusted initial state of new cards to display with unliked hearts.
- Rounded avatar images to match design specifications.
- Ensured universal functions/classes are suitable for all popups.
- Modified modal closing logic to handle successful server responses effectively, preserving input state and button text changes for error scenarios.

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Tyson, 

unfortunately I have to reject the project without reviewing:



* Please, place all **images** into folder `images` to structure the files
* When I add or remove cards, there are errors https://snipboard.io/OBnvfa.jpg 
* The same errors when I edit the profile data
* The avatar form is still not validated

ttruedell commented:
Addressed Feedback

- Modified form submission logic in the PopupWithForm and PopupWithConfirm to handle promise and account for synchronous and asynchronous cases
- Added form validator logic for the avatar edit modal
- Structured images/fonts, and modified avatar CSS style below 680px

gennady-bars commented on src/components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,52 @@
+    super.setEventListeners();
+    this._confirmButton.addEventListener("click", (event) => {
+      event.preventDefault();
+      this.setButtonText("Saving...");
```
`"Saving..."`

When you delete, you should set `Deleting...` in the button

gennady-bars commented on src/components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,52 @@
+      event.preventDefault();
+      this.setButtonText("Saving...");
+      this._handleConfirm();
+      const result = this._handleConfirm;
```
`      const result = this._handleConfirm;
`


Your code here doesn't work. Everything is in `index.js`

gennady-bars commented on src/components/PopupWithConfirm.js:
```diff
@@ -0,0 +1,52 @@
+
+  open() {
+    super.open();
+    this.resetButtonText();
```
You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -21,12 +36,33 @@ export default class PopupWithForm extends Popup {
       event.preventDefault();
       this._handleFormSubmit(this._getInputValues());
-      this.close();
+      const result = this._handleFormSubmit(this._getInputValues());
```
Please, don't duplicate submit handling

It should be in `index.js`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -21,12 +36,33 @@ export default class PopupWithForm extends Popup {
+
+  _isValidForm() {
+    const inputValues = this._getInputValues();
+    if (!inputValues.avatar) {
```
Please, delete `_isValidForm` because you check `avatar` here. Not all forms have avatar inputs

There should **not** be any code in universal functions that doesn’t suit every popup.

gennady-bars commented on src/pages/index.js:
```diff
@@ -23,88 +37,233 @@ const profileDescriptionInput = document.querySelector(
+  handleConfirm: () => {
+    if (currentCard) {
+      handleDeleteCard(currentCard);
+      deleteCardModal.close();
```
You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

gennady-bars reviewed (commented):
Everything is almost fine, but:

* I add **one** card, but 2 **new** cards appear on the page
* If the input is empty, the button should be disabled https://snipboard.io/rZQzJU.jpg 
* You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.
* There should **not** be any code in universal functions that doesn’t suit every popup.
* You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

ttruedell commented:
Follow-up Changes
I've addressed the feedback and made the following updates:

1. Promise Handling in PopupWithConfirm.js:

- Ensured that this._handleConfirm always returns a promise, eliminating the need for an else block in setEventListeners.
- Added a Promise.resolve() in the handleConfirm function to ensure the result is always a promise.

2. Button Text Handling:

- Updated the button text change logic to handle both success and failure cases within the finally block.
- Added detailed console logs to track the flow and ensure proper execution.

3. Form Submission:

- Corrected the return of the api.deleteCard method in handleDeleteCard to ensure it returns a promise.
- Updated the form submission logic to handle asynchronous operations correctly, preventing duplicate logs and ensuring proper button text reset.

4. Disabled Button Modifier for Avatar Modal:

- Added a disabled modifier and property to the button for the avatar modal to improve UX during loading states.

Please review the changes and let me know if there are any further adjustments needed.

gennady-bars commented on src/pages/index.js:
```diff
@@ -23,88 +37,225 @@ const profileDescriptionInput = document.querySelector(
+  handleFormSubmit: (data) => {
+    const userAvatar = {
+      avatar: data.avatar,
+    };
```
You need to call `renderLoading` when you submit the avatar to show the loading effect

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -20,13 +35,34 @@ export default class PopupWithForm extends Popup {
-      this.close();
+      // this._handleFormSubmit(this._getInputValues());
+      const result = this._handleFormSubmit(this._getInputValues());
+      if (result instanceof Promise) {
+        result
+          .then(() => {
+            this.close();
+          })
+          .catch((error) => {
+            console.error("Error during confirmation:", error);
+          });
+      } else {
+        this.close();
```
Please, delete the whole code from line 40 till 50 in `PopupWithForm`. It doesn't work. 


There should be only this


```js

  setEventListeners() {
    super.setEventListeners();
    this._form.addEventListener("submit", (event) => {
      event.preventDefault();
      this._handleFormSubmit(this._getInputValues());
    });
  }
```

And you don't need ` this.close();` here otherwise the modal closes immediately without waiting

gennady-bars commented on src/pages/index.js:
```diff
@@ -23,88 +37,225 @@ const profileDescriptionInput = document.querySelector(
+      .updateUserAvatar(userAvatar)
+      .then((updatedUserInfo) => {
+        userInfo.setUserInfo(updatedUserInfo);
+        editAvatarModal.close();
```
Please, call `toggleButtonState` for the avatar validator here to disable the button


```js

        editAvatarModal.close();  // first close 
        avatarFormValidator.toggleButtonState(); // only after closing the button will be disabled
```

* If the input is empty after submitting, the button should be disabled https://snipboard.io/3Bbk4y.jpg

gennady-bars reviewed (commented):
One more step is needed:


* If the input is empty after submitting, the button should be disabled https://snipboard.io/3Bbk4y.jpg 
* You need to call `renderLoading` when you submit the avatar to show the loading effect
* Please, delete the whole code from line 40 till 50 in `PopupWithForm`. It doesn't work.

ttruedell commented:
Follow-up Changes
I've made the following updates to address the recent feedback:

1. Disabled Button When Input is Empty:

- Added logic to disable the submit button if the input is empty after submission. This ensures the button remains disabled until valid input is provided.

2. Render Loading Effect for Avatar Submission:

- Implemented the renderLoading method call during avatar submission to show the loading effect.

3. Removed Non-Functional Code in PopupWithForm:

- Deleted the code from lines 40 to 50 in PopupWithForm.js as it was non-functional and unnecessary.

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
